src/shared/bootspec.c

```diff
@@ -1455,52 +1456,76 @@ static void print_addon(
-        printf("      cmdline: %s%s\n", special_glyph(SPECIAL_GLYPH_TREE_RIGHT), addon->cmdline);
+        printf("      options: %s%s\n", special_glyph(SPECIAL_GLYPH_TREE_RIGHT), addon->cmdline);
+}
+
+static int indent_embedded_newlines(char *cmdline, char **ret_cmdline) {
+        _cleanup_free_ char *t = NULL;
+        _cleanup_strv_free_ char **ts = NULL;
+
+        assert(ret_cmdline);
+        *ret_cmdline = NULL;
+
+        ts = strv_split_newlines(cmdline);
+        if (!ts)
+                return -1;
+
+        t = strv_join(ts, "\n              ");
+        if (!t)
+                return -1;
+
+        *ret_cmdline = TAKE_PTR(t);
+
+        return 0;
 }
 
 static int print_cmdline(
                 const BootEntry *e,
                 const BootEntryAddons *global_arr) {
 
-        _cleanup_free_ char *final_cmdline = NULL;
+        _cleanup_free_ char *options = NULL, *combined_cmdline = NULL, *t2 = NULL;
 
         assert(e);
 
         if (!strv_isempty(e->options)) {
-                _cleanup_free_ char *t = NULL, *t2 = NULL;
+                _cleanup_free_ char *t = NULL;
                 _cleanup_strv_free_ char **ts = NULL;
 
                 t = strv_join(e->options, " ");
                 if (!t)
                         return log_oom();
 
-                ts = strv_split_newlines(t);
-                if (!ts)
+                if (indent_embedded_newlines(t, &options) < 0)
                         return log_oom();
 
-                t2 = strv_join(ts, "\n              ");
+                printf("      options: %s\n", options);
+                t2 = strdup(options);
                 if (!t2)
                         return log_oom();
-
-                printf("  ukiCmdline: %s\n", t2);
-                final_cmdline = TAKE_PTR(t2);
         }
 
-        FOREACH_ARRAY(addon, global_arr->items, global_arr->count) {
-                print_addon(addon, "globalAddon");
-                if (!strextend(&final_cmdline, " ", addon->cmdline))
+        FOREACH_ARRAY(addon, global_arr->items, global_arr->n_items) {
+                print_addon(addon, "global-addon");
+                if (!strextend(&t2, " ", addon->cmdline))
                         return log_oom();
         }
 
-        FOREACH_ARRAY(addon, e->local_addons.items, e->local_addons.count) {
+        FOREACH_ARRAY(addon, e->local_addons.items, e->local_addons.n_items) {
                 /* Add space at the beginning of addon_str to align it correctly */
-                print_addon(addon, " localAddon");
-                if (!strextend(&final_cmdline, " ", addon->cmdline))
+                print_addon(addon, " local-addon");
+                if (!strextend(&t2, " ", addon->cmdline))
                         return log_oom();
         }
 
-        if (final_cmdline)
-                printf(" finalCmdline: %s\n", final_cmdline);
+        /* Don't print the combined cmdline if it's same as options. */
+        if (streq_ptr(t2, options))
```

Comment:
so you are comparing one of the strings which has the embedded newline indentation stuff applied with one which doesn't have that. please fix that, i.e. comapre both without the indentation stuff applied.

and just apply the indentation stuff right before display, because it's really only relevant for display.

Reply:
You are right, this actually compares a string that has the newline indentation stuff applied with a string that is the extension of the first string, where the first part still has the indentation stuff applied, but appended part has not. I misread this earlier, and assumed the 2nd string was not reindented at all.

There's the corner case you already mentioned though, regarding empty kernel cmdline add-ons, which I think we should cover though, simply because I expect that these add-ons will often be created programmatically and you never know whether people will always make sure to suppress the add-on if they figure out they don't need it...

Conceptually, I am pretty sure we should not operate on strings that are processed for display *except* if we use them for display. Hence, if you compare the two strings please make sure to compare the "raw" concatenations. It's conceptually cleaner, it's easy to do and it will address that corner case just fine.

Hence, when you display a command line, do the line break reindentation for it, display it, and then forget about it immediately, and continue with the original version. And if you then need to display an extended version, then apply the line break reindentation for the raw, concatenated version, display it, and fprget it too.

Reply:
Ok. It took me a little to figure the right function, but what I do now is checking if the cmdline section found in an addon contains WHITESPACE. If so, discard the addon at all.
I applied all the rest of the feedback.